File: ios-local/UTEDeadlineLocal/Services/MoodleURLValidator.swift
import Foundation

struct MoodleURLValidation {
    let ok: Bool
    let message: String
    let normalizedURL: String
}

enum MoodleURLValidator {
    private static let lmsHost = "utexlms.hcmute.edu.vn"
    private static let exportPath = "/calendar/export.php"
    private static let executePath = "/calendar/export_execute.php"

    static func validate(_ urlText: String) -> MoodleURLValidation {
        let trimmed = urlText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return MoodleURLValidation(ok: false, message: "Bạn chưa dán Calendar URL Moodle.", normalizedURL: "")
        }
        guard let url = URL(string: trimmed), let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return MoodleURLValidation(ok: false, message: "Link không đúng định dạng URL.", normalizedURL: "")
        }
        guard components.scheme == "https" else {
            return MoodleURLValidation(ok: false, message: "Link phải bắt đầu bằng https://", normalizedURL: "")
        }
        guard components.host == lmsHost else {
            return MoodleURLValidation(ok: false, message: "Link phải thuộc utexlms.hcmute.edu.vn.", normalizedURL: "")
        }
        if components.path == exportPath {
            return MoodleURLValidation(
                ok: false,
                message: "Đây mới là trang xuất lịch. Hãy copy Calendar URL dạng export_execute.php từ Moodle.",
                normalizedURL: ""
            )
        }
        guard components.path == executePath else {
            return MoodleURLValidation(ok: false, message: "Link phải là Calendar URL dạng /calendar/export_execute.php.", normalizedURL: "")
        }

        let query = components.queryItems ?? []
        guard query.first(where: { $0.name == "userid" })?.value?.nilIfBlank != nil else {
            return MoodleURLValidation(ok: false, message: "Link thiếu userid. Hãy copy lại đúng Calendar URL từ Moodle.", normalizedURL: "")
        }
        guard query.first(where: { $0.name == "authtoken" })?.value?.nilIfBlank != nil else {
            return MoodleURLValidation(ok: false, message: "Link thiếu authtoken. Hãy copy lại đúng Calendar URL từ Moodle.", normalizedURL: "")
        }

        return MoodleURLValidation(ok: true, message: "Calendar URL hợp lệ.", normalizedURL: trimmed)
    }

    static func mask(_ urlText: String) -> String {
        guard let components = URLComponents(string: urlText.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            return "Calendar URL đã lưu"
        }
        let userId = components.queryItems?.first(where: { $0.name == "userid" })?.value?.nilIfBlank ?? "..."
        let preset = components.queryItems?.first(where: { $0.name == "preset_time" })?.value?.nilIfBlank ?? "..."
        return "\(lmsHost) - userid=\(userId) - token=*** - \(preset)"
    }
}

File: android/app/src/main/java/com/utex/deadline/sync/MoodleUrlValidator.kt
package com.utex.deadline

import android.net.Uri

object MoodleUrlValidator {
    private const val LMS_HOST = "utexlms.hcmute.edu.vn"
    private const val EXPORT_PATH = "/calendar/export.php"
    private const val EXECUTE_PATH = "/calendar/export_execute.php"

    fun validate(urlText: String): ValidationResult {
        val trimmed = urlText.trim()
        if (trimmed.isBlank()) {
            return ValidationResult(false, "Bạn chưa dán Calendar URL Moodle.")
        }

        val uri = try {
            Uri.parse(trimmed)
        } catch (_: Exception) {
            return ValidationResult(false, "Link không đúng định dạng URL.")
        }

        if (uri.scheme != "https") {
            return ValidationResult(false, "Link phải bắt đầu bằng https://")
        }
        if (uri.host != LMS_HOST) {
            return ValidationResult(false, "Link phải thuộc utexlms.hcmute.edu.vn.")
        }
        if (uri.path == EXPORT_PATH) {
            return ValidationResult(false, "Đây mới là trang xuất lịch. Hãy bấm Lấy địa chỉ mạng của lịch trên Moodle rồi copy Calendar URL.")
        }
        if (uri.path != EXECUTE_PATH) {
            return ValidationResult(false, "Link phải là Calendar URL dạng /calendar/export_execute.php.")
        }
        if (uri.getQueryParameter("userid").isNullOrBlank()) {
            return ValidationResult(false, "Link thiếu userid. Hãy copy lại đúng Calendar URL từ Moodle.")
        }
        if (uri.getQueryParameter("authtoken").isNullOrBlank()) {
            return ValidationResult(false, "Link thiếu authtoken. Hãy copy lại đúng Calendar URL từ Moodle.")
        }

        return ValidationResult(true, "Calendar URL hợp lệ.", trimmed)
    }

    fun mask(urlText: String): String {
        val uri = try {
            Uri.parse(urlText.trim())
        } catch (_: Exception) {
            return "Calendar URL đã lưu"
        }
        val userId = uri.getQueryParameter("userid").orEmpty().ifBlank { "..." }
        val preset = uri.getQueryParameter("preset_time").orEmpty().ifBlank { "..." }
        return "utexlms.hcmute.edu.vn - userid=$userId - token=*** - $preset"
    }

    data class ValidationResult(
        val ok: Boolean,
        val message: String,
        val normalizedUrl: String = ""
    )
}
